owner-onboarding-server/src/ov_reply.rs

```diff
@@ -0,0 +1,85 @@
+
+
+use fdo_data_formats::{ownershipvoucher::OwnershipVoucher};
+use serde::{Serialize, Deserialize};
+
+use warp::{hyper::StatusCode, http::HeaderValue, Buf};
+
+
+#[derive(Clone, Copy)]
+#[derive(Serialize,Deserialize)]
+enum ErrorCode {
+    IncompleteVoucher,
+    ParseError,
+    InvalidNumberOfVouchers,
+    UnwonedVoucher,
+}
+
+#[derive(Serialize,Deserialize)]
+struct Error<T:Serialize> {
+    error_code: Option<ErrorCode>,
+    error_details : Option<ErrorDetails<T>>
+}
+#[derive(Serialize,Deserialize)]
+struct ErrorDetails<T:Serialize >  {
+    key : Option<ErrorCode>,
+    value: Option<T>
+}
+
+trait Response<T> {
+    fn generate(e: ErrorCode,value: T ) -> Self;
+}
+
+impl<T> Response<T> for Error<T> where T: Serialize {
+    fn generate(e: ErrorCode,value: T) -> Self {
+              Error {
+                error_code:  Some(e),
+                error_details: Some(ErrorDetails{
+                    key : Some(e),
+                    value: Some(value),
+                })
+            }
+        
+    }
+}
+
+
+pub(crate) async fn add_ov(
+    c: HeaderValue, 
+    body: impl Buf,
+    user_data: super::OwnerServiceUDT,
+) -> Result<impl warp::Reply,warp::Rejection> 
+{
+        let c = c.to_str().unwrap().parse::<usize>().unwrap();
+        let ovs = OwnershipVoucher::many_from_pem(body.chunk());
+
+        OwnershipVoucher::pem_read_error_count(body.chunk());
+        if ovs.is_err() {
+            let  e =  Error::generate(ErrorCode::ParseError, "Parse Error");
+            let e = warp::reply::json(&e);
+            return Ok(warp::reply::with_status(e, StatusCode::BAD_REQUEST));
+        }
+        
+        let ovs = ovs.unwrap();
+        if ovs.len() != c {
+                let  e =  Error::generate(ErrorCode::IncompleteVoucher, ovs.len());
+                let e = warp::reply::json(&e);
+                return Ok(warp::reply::with_status(e, StatusCode::BAD_REQUEST));
+        }  
+        
+        
+        
+        // When every check passes
+        let uuid: Vec<String> = ovs.iter()
+                .map(|ov| {
+                   let a = ov.clone();
+                    user_data
+                    .ownership_voucher_store
+                    .store_data(a.header().guid().clone(), a);
```

Comment:
Its not saving ov
@puiterwijk

Reply:
Right, because this is a future.
You will need to make sure to await its results, and the compiler should be warning you about this, I believe?